File: 백준/Gold/14499. 주사위 굴리기/주사위 굴리기.py
#14499
N, M, r, c, K = map(int,input().split())
if M>1:
    marr = [list(map(int,input().split())) for _ in range(N)]
else:
    marr = [[int(input())] for _ in range(N)]
if K>1:
    commands = list(map(int,input().split()))
else:
    commands = [int(input())]

dice = [0]*7

def left():
    dice[4],dice[1],dice[3], dice[6] = dice[1], dice[3], dice[6], dice[4]

def right():
    dice[4],dice[1],dice[3], dice[6] = dice[6],dice[4],dice[1], dice[3]

def up():
    dice[2],dice[6] = dice[6], dice[2]
    dice[1], dice[5] = dice[5], dice[1]
    dice[2],dice[5] = dice[5], dice[2]

def down():
    dice[2],dice[6] = dice[6], dice[2]
    dice[1], dice[5] = dice[5], dice[1]
    dice[1],dice[6] = dice[6], dice[1]

for com in commands:
    if com==1:
        if c+1>=M:
            continue
        right()
        c+=1
    elif com==2:
        if c-1<0:
            continue
        left()
        c-=1
    elif com==3:
        if r-1<0:
            continue
        up()
        r-=1
    elif com==4:
        if r+1>=N:
            continue
        down()
        r+=1
    if marr[r][c]==0:
        marr[r][c]=dice[6]
    else:
        dice[6]=marr[r][c]
        marr[r][c]=0
    print(dice[1])

File: 백준/Gold/14499. 주사위 굴리기/주사위 굴리기.cc
#include <iostream>
using namespace std;
struct _dice{
    int bot;
    int top;
    int up;
    int down;
    int left;
    int right;
} dice;
struct _dicePos{
    int r;
    int c;
} dicePos;
int N,M;
int grid[20][20];




void right(_dice &d){
    int tL, tB, tR, tT;
    tL=d.bot;
    tB=d.right;
    tR=d.top;
    tT=d.left;
    d.left=tL;
    d.bot=tB;
    d.right=tR;
    d.top=tT;
}

void left(_dice &d){
    int tL, tB, tR, tT;
    tL=d.top;
    tB=d.left;
    tR=d.bot;
    tT=d.right;
    d.left=tL;
    d.bot=tB;
    d.right=tR;
    d.top=tT;
}

void up(_dice &d){
    int tU,tB,tD,tT;
    tU = d.top;
    tB = d.up;
    tD = d.bot;
    tT = d.down;
    d.top = tT;
    d.bot = tB;
    d.up = tU;
    d.down = tD;
}

void down(_dice &d){
    int tU,tB,tD,tT;
    tU = d.bot;
    tB = d.down;
    tD = d.top;
    tT = d.up;
    d.top = tT;
    d.bot = tB;
    d.up = tU;
    d.down = tD;
}

void botWork(_dice &d, _dicePos &p){
        if (grid[p.r][p.c]==0){
            grid[p.r][p.c]=d.bot;
        }
        else{
            d.bot = grid[p.r][p.c];
            grid[p.r][p.c]=0;
        }
}

int main(){
    int K;
    cin>>N>>M>>dicePos.r>>dicePos.c>>K;
    for(int i=0;i<N;i++){
        for (int j=0;j<M;j++){
            cin>>grid[i][j];
        }
    }
    dice.top = 0;
    dice.left=0;
    dice.right=0;
    dice.up=0;
    dice.down=0;
    dice.bot=0;


    int commands[1000];
    for(int i=0;i<K;i++){
        cin>>commands[i];
    }

    
    
    for(int i=0;i<K;i++){
        int command = commands[i];
        switch(command){
            case 1:
                if (dicePos.c+1<M){
                    right(dice);
                    dicePos.c+=1;
                    botWork(dice,dicePos);
                    cout<<dice.top<<'\n';
                }
                break;
            case 2:
                if (dicePos.c-1>=0){
                    left(dice);
                    dicePos.c-=1;
                    botWork(dice,dicePos);
                    cout<<dice.top<<'\n';
                }
                break;
            case 3:
                if (dicePos.r-1>=0){
                    up(dice);
                    dicePos.r-=1;
                    botWork(dice,dicePos);
                    cout<<dice.top<<'\n';
                }
            break;
            case 4:
                if (dicePos.r+1<N){
                    down(dice);
                    dicePos.r+=1;
                    botWork(dice,dicePos);
                    cout<<dice.top<<'\n';
                
                }            
                break;
            default:
            break;
        }
    }
}